libraries/base/Data/Semigroup.hs

```diff
@@ -18,21 +18,37 @@
 -- Maintainer  :  libraries@haskell.org
 -- Stability   :  provisional
 -- Portability :  portable
+-- 
+-- A type @a@ is a 'Semigroup' if it provides an associative function ('<>')
+-- that lets you combine any two values of type @a@ into one. 
 --
--- In mathematics, a semigroup is an algebraic structure consisting of a
--- set together with an associative binary operation. A semigroup
--- generalizes a monoid in that there might not exist an identity
--- element. It also (originally) generalized a group (a monoid with all
--- inverses) to a type where every element did not have to have an inverse,
--- thus the name semigroup.
+-- ==== __Examples__
+-- >>> sconcat (1 :| [2, 3, 4]) :: Min Int
+-- Min {getMin = 1}
+-- >>> sconcat (1 :| [2, 3, 4]) :: Max Int
+-- Max {getMax = 4}
+-- >>> "foo" <> "bar"
+-- "foobar"
+-- >>> Last 1 <> Last 2 <> Last 3
+-- Last {getLast 3}
+-- >>> First 1 <> First 2 <> First 3
+-- First {getFirst = 1}
 --
--- The use of @(\<\>)@ in this module conflicts with an operator with the same
+-- Note that the use of @(\<\>)@ in this module conflicts with an operator with the same
 -- name that is being exported by Data.Monoid. However, this package
 -- re-exports (most of) the contents of Data.Monoid, so to use semigroups
 -- and monoids in the same package just
 --
 -- > import Data.Semigroup
 --
+-- ==== __Theoretical Context__
+--
+-- A 'Semigroup' is a generalization of a 'Monoid'. Unlike the 'Semigroup', the 'Monoid'
+-- requires the presence of a neutral element ('mempty') in addition to the associative
+-- operator. The neutral element furthermore lifts the restriction of 'Semigroup' to work 
+-- only on non-empty types (as can be seen by the use of the NonEmpty list type instead 
+-- of a normal list).
```

Comment:
OK, this wording is more clear, but now I think it is wrong :o Semigroup is not restricted to work only on non-empty types. Since Semigroup is a generaliation of Monoid, all Monoids are semigroups, but there are some semigroups which cannot provide a lawful `mempty` value and have a Monoid instance.

The requirement for a neutral element doesn't lift a restriction; it *incurs* a restriction on `Monoid`al types. I think I would word this as:

> The requirement for a neutral element forbids many types from being a full `Monoid`, like `NonEmpty`.